model/incoming_webhook.go

```diff
@@ -63,11 +98,201 @@ func IncomingWebhookListFromJson(data io.Reader) []*IncomingWebhook {
+			return false
+		}
+	} else {
+		if len(idxValue) > 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (s StringInterface) ValidHeaderModel(mapKey string) (string, bool) {
+	var isSet bool = false
+	var val string = ""
+	if s[mapKey] != nil {
+		isSet = true
+	}
+	if headerPart,ok := s[mapKey].(string); ok {
+		val = headerPart
+	}
+	return val, isSet
+} 
+
+func (o *IncomingWebhook) IsValidModels() bool {
+	if ! (o.HmacAlgorithm == "HMAC-SHA1" || o.HmacAlgorithm == "HMAC-SHA256") {
+		return false
+	}
+
+	if ! (len(o.SignedContentModel) > 0 && len(o.SignedContentModel) < 6) {
+		return false
+	}
+
+	hmacModelHeader, ok := o.HmacModel["HeaderName"].(string)
+	if  ! (ok && 0 < len(hmacModelHeader) && len(hmacModelHeader) < 26) {
+		return false
+	}
+	if ! ValidModelSplitter(&o.HmacModel) {
+		return false
+	}
+
+	if pre,pSet := o.HmacModel.ValidHeaderModel("Prefix"); pSet && pre == "" {
+		return false
+	}
+
+	timestampModelH := o.TimestampModel["HeaderName"]
+	if timestampModelH == nil {
+		o.TimestampModel = StringInterface{}
+	} else {
+		tsModelHeader, ok1 := timestampModelH.(string)
+		if  ! (ok1 && 0 < len(tsModelHeader) && len(tsModelHeader) < 26) {
+			return false
+
+		}
+
+		if ! ValidModelSplitter(&o.TimestampModel) {
+			return false
+		}
+
+		if pre,pSet := o.TimestampModel.ValidHeaderModel("Prefix"); pSet && (pre == "" || len(pre) > 8) {
+			return false
+		}
+		
+	}
+
+	return true
+}
+
+// this should be replaced with a better way
+func ParseHeaderModel(signingPart StringInterface) (HeaderModel, *AppError) {
+	// var validate *validator.Validate
+	jsonSignPart, err := json.Marshal(signingPart)
+
+	if err != nil {
+		return HeaderModel{},NewAppError("IncomingWebhook.IsValid", "model.incoming_hook.invalid_signing_part1.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	headerModel := HeaderModel{}
+	if err = json.Unmarshal(jsonSignPart, &headerModel); err != nil {
+		// do error check
+		return HeaderModel{},NewAppError("IncomingWebhook.IsValid", "model.incoming_hook.invalid_signing_part2.app_error", nil, "", http.StatusBadRequest)
+	}
+	return headerModel, nil
+}
+
+func (sp SignedIncomingHook) VerifySignature(secretToken string) bool {
+	var mac hash.Hash
+	if sp.Algorithm == "HMAC-SHA1" {
+		mac = hmac.New(sha1.New, []byte(secretToken))
+	} else {
+		mac = hmac.New(sha256.New, []byte(secretToken))
+	}
+	
+	mac.Write(*sp.ContentToSign)
+	sha := mac.Sum(nil)
+
+	sig,err := hex.DecodeString(sp.Signature)
+	if err != nil {
+		return false
+	}
+
+	return hmac.Equal(sig, sha)
+}
+
+func (o *IncomingWebhook) ParseHeader(signedHook *SignedIncomingHook, r *http.Request) *AppError {
+	timestampModel,err := ParseHeaderModel(o.TimestampModel)
+	if err != nil {
+		return err
+	}
+
+	digestModel,err := ParseHeaderModel(o.HmacModel)
+	if err != nil {
+		return err
+	}
+
+	time_stamp:= new(string)
+	digest := new(string)
+	for k,v := range map[*string]HeaderModel{time_stamp:timestampModel, digest:digestModel} {
```

Comment:
I am not sure what this for is doing. Could you explain a bit further?

Reply:
@larkox This uses the models provided incoming webhook creator to parse the expected headers. For example HeaderName and then SplitBy, Index and Prefix and finally getting the real digest and timestamp.

Reply:
So basically this code parses the timestamp and digest headers in webhook request to the format/model defined/expected by the webhook endpoint owner/admin. Both timestamp and digest headers are formatted in HeaderName, SplitBy, Index, Prefix. For example Github only uses digest formatted like X-Hub-Signature: Sha1:a812e6e6e6e6e6e6e6e6 and no timestamp. So as the endpoint owner on MM, I'd define Algorithm=HMAC-SHA1, Signature HeadName={X-Hub-Signature}Prefix(Sha1:). In this way, this code finds the pure signature a812e6e6e6e6e6e6e6e6 after one loop.